### DISCUSSION TITLE
IPython console when running Python code interactively with REPL

### DISCUSSION THREAD

**Discussion by yalap13** [zed-industries/zed#23370] [Feature Requests]
### Check for existing issues

- [x] Completed

### Describe the feature

Similar to what exists in Spyder, it would be nice to have access to the IPython console when running python code interactively using REPL to allow running a file and running additional commands. There could be a setting to automatically open or not the console whenever a line or cell of python code is ran interactively.

### Environment

Zed: v0.165.4 (Zed)
OS: Linux Wayland fedora 41
Memory: 15.3 GiB
Architecture: x86_64
GPU: Intel(R) Graphics (ADL GT2) || Intel open-source Mesa driver || Mesa 24.2.8

### If applicable, add mockups / screenshots to help present your vision of the feature

Here is what the console looks like in Spyder :
![spyder_screenshot](https://github.com/user-attachments/assets/636ef04c-f18d-41b1-ae58-c8c590a26ace)

**Comment by AlexisMrl** [comment]
I also believe it would be of great use. I often work in python and the ability to executes cells from a file is great but I often want to run one liners or arbitrary code that i do not want in my file. It's also nice to not have the cell outputs to be inside the editor area.

For what it's worth, it's currently kind of possible, by locating the kernel file with the magic `%connect_info` and then opening a jupyter qtconsole connected to the kernel with: `jupyter qtconsole --existing /run/<...>/jupyter/kernel-zed-<num>.json`.
I did not find a way to open an ipython console connected to a kernel (rather than a qtconsole, which is a standalone gui app).

Of course it should also work with other REPL too, that probably have different ways of connecting to existing kernels.

**Reply by moritzwilksch** [reply]
Seconded! VSCode's "Interactive Window" is a similar feature I love and currently miss in Zed. It runs an ipykernel and integrates seamlessly into the IDE, supporting features like plot display. It also detects cell markers in .py files ("# %%" or "#%%") automatically, allowing you to run an entire cell (lines between two markers) with a keyboard shortcut.

**Comment by vultix** [comment]
This is the one feature preventing me from switching g from Pycharm. Like vscode, it runs an ipykernel and is integrated with the IDE, giving a gui for auto completions, supporting plots, running code cell blocks, and inspecting variables.

**Comment by shner-elmo** [comment]
+1
This would be a great feature if it had syntax highlighting and code completion.
Basically an `IPython` console

**Comment by giorgioskij** [comment]
+1
this feature is a main part of my workflow as a ml researcher

**Reply by mashu** [reply]
> +1 this feature is a main part of my workflow as a ml researcher

Same here, I come from #16024 
Being able to see what's sent and switch between REPL and inline edits is needed.

Did anyone have success with it?

**Comment by vovavili** [comment]
This, side-by-side Git and a DataFrame explorer are the only three missing features that still keep me in PyCharm at this point. Having this in Zed would be beyond awesome - while I am fine with typing in `ipython` in the console, PyCharm's variable explorer makes REPL experience much more seamless.
